specs/APIReview_TextureStream.md

```diff
@@ -0,0 +1,592 @@
+  /// The API also registers created shared handle to the browser once it
+  /// created the resource.
+  HRESULT CreateBuffer(
+    [in] UINT32 width,
+    [in] UINT32 height,
+    [out, retval] ICoreWebView2StagingTexture** buffer);
+  /// GetAvailableBuffer can be called on any thread like SetBuffer.
+  HRESULT GetAvailableBuffer([out, retval] ICoreWebView2StagingTexture** buffer);
+  /// Remove texture buffer when the host removes the backed 2D texture.
+  /// The host can save the existing resources by deleting 2D textures
+  /// when it changes the frame sizes.
+  HRESULT RemoveBuffer([in] ICoreWebView2StagingTexture* buffer);
+  /// Indicates that the buffer is ready to present.
+  /// The buffer must be retrieved from the GetAvailableBuffer.
+  /// The host writes new texture to the local shared 2D texture of
+  /// the buffer id, which is created via CreateBuffer.
+  /// SetBuffer API can be called in any thread.
+  HRESULT SetBuffer([in] ICoreWebView2StagingTexture* buffer,
+    [in] ULONGLONG timestamp);
+  /// Render texture that is current set ICoreWebView2StagingTexture.
+  HRESULT Present();
+  /// Stop streaming of the current stream id.
+  /// API calls of Present, CreateBuffer will fail after this
+  /// with an error of COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED.
+  /// The Javascript can restart the stream with getTextureStream.
+  HRESULT Stop();
+  /// Event handler for those that occur at the Renderer side, the example
+  /// are CreateBuffer, Present, or Stop.
+  HRESULT add_TextureError(
+      [in] ICoreWebView2StagingTextureStreamTextureErrorEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+  /// Remove listener for texture error event.
+  HRESULT remove_TextureError([in] EventRegistrationToken token);
+  /// Updates d3d Device when it is updated by RenderAdapterLUIDChanged
+  /// event.
+  HRESULT UpdateD3DDevice([in] IUnknown* d3dDevice);
+  /// Event handler for receiving texture by Javascript.
+  /// There is no Start event for web texture. Whenever texture are sent
+  /// from the Javascript, the event is triggered.
+  HRESULT add_WebTextureReceived(
+      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+  HRESULT remove_WebTextureReceived([in] EventRegistrationToken token);
+
+  /// Event handler for stopping of the receiving texture stream.
+  HRESULT add_WebTextureStreamStopped(
+      [in] ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+  HRESULT remove_WebTextureStreamStopped([in] EventRegistrationToken token);
+
+  /// Adds an allowed url origin for the given stream id for web texture
+  /// operation.  Javascript can send texture stream to the host only when
+  /// the origin it runs are allowed by the host.
+  HRESULT AddWebTextureAllowedOrigin([in] LPCWSTR origin);

```

Comment:
Why do we have to have a separate allow list for this? Can we just use the one AddAllowedOrigin? 

If we do need it separate how about modifying AddAllowedOrigin to take a permission parameter. I guess its not possible to have an origin that's not allowed to receive textures but is allowed to send them. 
So you can do 

AddAllowedOrigin("https://example.com", true) where the second param is a bool if they're allowed to send textures back.

Reply:
It is requested API from the Teams because of their security review board. They prefer independent APIs for respective get and register texture stream case, which is what Electron based solution has it.

Reply:
We can still have the separate features with separate permissions, but one method AddAllowedOrigin where you specify the permission per origin. This lets us ensure that you can't get into a weird state of an origin allowed to send textures but not allowed to receive textures. It should also make it easier to read and audit code to see what origins get what permissions.

Reply:
It might cause more confusion to the developer whether they need to add false or true when calling RemoveAllowedOrigin? for example, remove bidi only, or remove both of them, (potentially normal direction only). 

For normal direction only, there was a request to allow registerTextureStream without getTextureStream a while ago just in case of future use from the Teams, but we rejected it at that time as it isn't supported in the current Electron app.  I'm afraid though if we have single API for both directions, we might have to create API again if the request becomes a product request in future.

We expect that most common scenario is normal direction only, only certain scenario needs bidi (registerTextureStream).

However, I'm open to change though, please let me know what you think.

Reply:
I've talked with Teams team about this and they are fine with single API, I will combine these two into single AddAlloedOrigin.

Reply:
Change as suggested.